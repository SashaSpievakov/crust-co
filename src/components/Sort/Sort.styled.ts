import styled, { css } from 'styled-components';
import { BsCaretDownFill, BsCaretUpFill } from 'react-icons/bs';

interface PopupProps {
  open: boolean;
}

interface LiProps {
  chosen: boolean;
}

const mixinArrow = css`
  font-size: 15px;
  margin-right: 5px;
`;

const mixinBackground = css`
  background-color: ${({ theme }) => theme.colors.secondaryTransparent};
`;

const Wrapper = styled.section`
  width: 127px;
  position: relative;
  margin-top: 15px;
  cursor: pointer;

  @media screen and (max-width: 650px) {
    align-self: end;
  }
`;

const Label = styled.div`
  display: flex;
  align-items: center;

  b {
    margin-right: 8px;
  }

  span {
    color: ${({ theme }) => theme.colors.primary};
    border-bottom: 1px dashed ${({ theme }) => theme.colors.primary};
  }
`;

const ArrowUp = styled(BsCaretUpFill)`
  ${mixinArrow}
`;

const ArrowDown = styled(BsCaretDownFill)`
  ${mixinArrow}
`;

const Popup = styled.section<PopupProps>`
  visibility: hidden;
  position: absolute;
  right: -10px;
  top: 0px;
  width: 160px;
  margin-top: 15px;
  padding: 10px 0;
  background: ${({ theme }) => theme.colors.fourth};
  box-shadow: 0px 5px 15px rgba(0, 0, 0, 0.09);
  border-radius: 10px;
  opacity: 0;
  cursor: auto;
  z-index: 1;
  transition: 0.25s;

  ${({ open }) =>
    open &&
    css`
      visibility: visible;
      top: 22.5px;
      opacity: 1;
    `}
`;

const Li = styled.li<LiProps>`
  padding: 12px 20px;
  cursor: pointer;

  &:hover {
    ${mixinBackground}
  }

  ${({ chosen }) =>
    chosen &&
    css`
      font-weight: bold;
      color: ${({ theme }) => theme.colors.primary};
      ${mixinBackground}
    `}
`;

export { Wrapper, Label, ArrowUp, ArrowDown, Popup, Li };
