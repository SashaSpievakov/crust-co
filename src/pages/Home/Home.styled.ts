import styled from 'styled-components';

const Title = styled.h2`
  margin: 35px 0;

  @media screen and (max-width: 650px) {
    margin: 20px 0;
  }
`;

const Top = styled.section`
  display: flex;
  justify-content: space-between;
  align-items: center;
  margin-bottom: 30px;

  @media screen and (max-width: 900px) {
    flex-direction: column;
    align-items: start;
    margin-bottom: 15px;
  }
`;

const Block = styled(Top)`
  flex-direction: row;
  align-items: center;

  @media screen and (max-width: 900px) {
    margin-bottom: 50px;
  }

  @media screen and (max-width: 650px) {
    flex-direction: column;
    align-items: start;
  }
`;

const ErrorHeading = styled.h2`
  font-size: 32px;
  font-weight: 700;
  margin-bottom: 10px;

  @media screen and (max-width: 450px) {
    font-size: 25px;
  }
`;

const ErrorParagraph = styled.p`
  font-size: 22px;

  @media screen and (max-width: 650px) {
    font-size: 18px;
  }

  @media screen and (max-width: 450px) {
    font-size: 16px;
  }
`;

export { Title, Top, Block, ErrorHeading, ErrorParagraph };
