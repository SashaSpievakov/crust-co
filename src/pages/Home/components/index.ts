export { Categories } from './Categories';
export { ProductsContainer } from './ProductsContainer';
