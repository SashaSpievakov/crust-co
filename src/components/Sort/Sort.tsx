import { useEffect, useState, useRef, memo } from 'react';
import { useAppDispatch, useAppSelector } from '../../hooks/reduxHooks';

import { selectSort, setSort } from '../../store/slices/sortSlice';
import { Wrapper, Label, ArrowUp, ArrowDown, Popup, Li } from './Sort.styled';

interface CategoriesProps {
  sortNamesArr: string[];
}

const Sort = memo(({ sortNamesArr }: CategoriesProps) => {
  const activeSort = useAppSelector(selectSort);
  const [open, setOpen] = useState(false);
  const dispatch = useAppDispatch();
  const sortRef = useRef<HTMLDivElement>(null);

  const chosenSortName = sortNamesArr[activeSort];

  const changeActiveSortName = (index: number) => {
    dispatch(setSort(index));
    setOpen(false);
  };

  useEffect(() => {
    const handleBodyClick = (e: MouseEvent) => {
      if (sortRef.current && !e.composedPath().includes(sortRef.current)) {
        setOpen(false);
      }
    };
    if (open) {
      document.body.addEventListener('click', handleBodyClick);
    }

    return () => {
      document.body.removeEventListener('click', handleBodyClick);
    };
  }, [open]);

  return (
    <Wrapper ref={sortRef}>
      <Label onClick={() => setOpen(!open)}>
        {open ? (
          <ArrowUp data-testid="popupIcon" />
        ) : (
          <ArrowDown data-testid="popupIcon" />
        )}
        <b>Sort by</b>
        <span data-testid="currenChosenSortName">{chosenSortName}</span>
      </Label>
      <Popup open={open} aria-hidden={!open}>
        <ul aria-hidden={!open} data-testid="popupList">
          {sortNamesArr.map((sortName, i: number) => (
            <Li
              key={sortName}
              onClick={() => changeActiveSortName(i)}
              chosen={chosenSortName === sortName}
            >
              {sortName}
            </Li>
          ))}
        </ul>
      </Popup>
    </Wrapper>
  );
});

export default Sort;
