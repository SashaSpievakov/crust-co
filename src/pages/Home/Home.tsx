import { useEffect, useState } from 'react';

import { Title, Block, Top, Error } from '../../styles/Base.styled';
import { selectSort } from '../../store/slices/sortSlice';
import { selectCategory } from '../../store/slices/categorySlice';
import Categories from '../../components/Categories/Categories';
import Sort from '../../components/Sort/Sort';
import SearchItems from '../../components/SearchItems/SearchItems';
import { IPizzaItem } from '../../models/IPizzaItem';
import { useAppSelector } from '../../hooks/reduxHooks';
import pizzasAPI from '../../services/PizzasService';
import ProductsContainer from '../../components/ProductsContainer/ProductsContainer';

export const sortNamesArr: string[] = ['rating', 'price', 'A to Z'];

const Home = () => {
  const [pizzas, setPizzas] = useState<IPizzaItem[]>([]);
  const activeCategory = useAppSelector(selectCategory);
  const activeSort = useAppSelector(selectSort);

  const sortedActiveName = sortNamesArr[activeSort];

  const sortPropertyName = (property: string) => {
    if (property === 'A to Z') {
      return 'name&order=asc';
    }
    return `${property}&order=desc`;
  };

  const sortedPropertyName = sortPropertyName(sortedActiveName);

  const { data, isSuccess, isError, isLoading } = pizzasAPI.useFetchPizzasQuery(
    {
      activeCategory,
      sortedPropertyName,
    },
  );

  useEffect(() => {
    if (isSuccess) {
      setPizzas(data);
    }
  }, [activeCategory, sortedPropertyName, data, isSuccess]);

  return (
    <>
      <Top>
        <Categories />
        <Sort sortNamesArr={sortNamesArr} />
      </Top>
      <Block>
        <Title>All Pizzas</Title>
        <SearchItems />
      </Block>
      {isError ? (
        <Error>
          <h2>Request Error</h2>
          <p>Coudn&apos;t get store items. Try your request again later.</p>
        </Error>
      ) : (
        <ProductsContainer isLoading={isLoading} items={pizzas} />
      )}
    </>
  );
};

export default Home;
