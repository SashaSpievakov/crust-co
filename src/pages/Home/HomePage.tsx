import { ChangeEvent, FC, useEffect, useState } from 'react';
import { useSearchParams } from 'react-router-dom';
import { ErrorRequest } from 'src/components';
import { useDebounce } from 'src/hooks';
import { Container } from 'src/styles/Base.styled';

import { DropdownSelect, SearchItems } from '../../components/UI';
import { IPizzaItem } from '../../models';
import { pizzasAPI } from '../../services';
import { modifySearchParamsName } from '../../utils/modifySearchParamsName';
import { Categories, ProductsContainer } from './components';
import { Block, Title, Top } from './HomePage.styled';

export const sortNamesArr: string[] = ['rating', 'price', 'A to Z'];

export const HomePage: FC = () => {
  const [searchParams, setSearchParams] = useSearchParams();
  const [pizzas, setPizzas] = useState<IPizzaItem[]>([]);
  const [searchValue, setSearchValue] = useState<string>('');
  const debouncedSearchValue = useDebounce(searchValue, 100);
  const [activeSort, setActiveSort] = useState<number>(
    parseInt(searchParams.get('sortBy') || '0', 10) || 0,
  );
  const [activeCategory, setActiveCategory] = useState<number>(
    parseInt(searchParams.get('category') || '0', 10) || 0,
  );

  const sortSearchParam = modifySearchParamsName(sortNamesArr[activeSort]);
  const { data, isSuccess, isError, isLoading } = pizzasAPI.useFetchPizzasQuery(
    {
      activeCategory,
      sortSearchParam,
    },
  );

  useEffect(() => {
    document.title = 'Crust & Co.';
  }, []);

  useEffect(() => {
    if (isSuccess) {
      setPizzas(data);
    }
  }, [activeCategory, sortSearchParam, data, isSuccess]);

  const handleSortSelect = (index: number): void => {
    setActiveSort(index);
    setSearchParams({
      sortBy: index.toString(),
      category: activeCategory.toString(),
    });
  };

  const handleCategorySelect = (index: number): void => {
    setActiveCategory(index);
    setSearchParams({
      category: index.toString(),
      sortBy: activeSort.toString(),
    });
  };

  const onChangeSearch = (e: ChangeEvent<HTMLInputElement>): void => {
    setSearchValue(e.target.value);
  };

  const onClickClose = () => {
    setSearchValue('');
  };

  return isError ? (
    <ErrorRequest />
  ) : (
    <Container data-testid="homePage">
      <Top>
        <Categories
          activeCategory={activeCategory}
          onSelect={handleCategorySelect}
        />
        <DropdownSelect
          chosenValue={sortNamesArr[activeSort]}
          sortNamesArr={sortNamesArr}
          onSelect={handleSortSelect}
        />
      </Top>
      <Block>
        <Title>All Pizzas</Title>
        <SearchItems
          value={searchValue}
          width="240px"
          onChange={onChangeSearch}
          onClickClose={onClickClose}
          expandedWidth="300px"
        />
      </Block>
      <ProductsContainer
        isLoading={isLoading}
        items={pizzas}
        searchValue={debouncedSearchValue}
      />
    </Container>
  );
};
