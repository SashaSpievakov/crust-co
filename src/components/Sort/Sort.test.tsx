import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { sortNamesArr } from '../../pages/Home/Home';
import rendererWithProviders from '../../tests/helpers/rendererWithProviders';
import renderWithProviders from '../../tests/helpers/renderWithProviders';
import Sort from './Sort';

describe('Sort Test', () => {
  test('renders the Sort component', () => {
    const snapshot = rendererWithProviders(
      <Sort sortNamesArr={sortNamesArr} />,
    );
    expect(snapshot).toMatchSnapshot();
  });

  describe('tests general functionality', () => {
    beforeEach(() => {
      renderWithProviders(<Sort sortNamesArr={sortNamesArr} />);
    });

    test('changes sort state in redux', () => {
      const arrow = screen.getByTestId('popupIcon');
      const chosenSornName = screen.getByTestId('currenChosenSortName');

      userEvent.click(arrow);
      userEvent.click(screen.getByText(/a to z/i));

      expect(chosenSornName).toHaveTextContent(/a to z/i);
    });

    test('opens the popup', () => {
      const arrow = screen.getByTestId('popupIcon');
      const list = screen.getByRole('list', { hidden: true });

      userEvent.click(arrow);

      expect(list).toBeVisible();
    });

    test('closes the popup on icon click', () => {
      const arrow = screen.getByTestId('popupIcon');
      const list = screen.getByRole('list', { hidden: true });

      userEvent.click(arrow);
      userEvent.click(screen.getByTestId('popupIcon'));

      expect(list).not.toBeVisible();
    });

    test('closes the popup on list item click', () => {
      const sortCaption = screen.getByText(/sort by/i);
      const list = screen.getByRole('list', { hidden: true });

      userEvent.click(sortCaption);
      userEvent.click(screen.getByText(/price/i));

      expect(list).not.toBeVisible();
    });
  });

  describe('test integration with the whole appliation', () => {
    test('closes popup on body click', () => {
      renderWithProviders(null, true);
      const sortCaption = screen.getByText(/sort by/i);
      const bodyHeading = screen.getByRole('heading', { level: 2 });
      const list = screen.getByTestId('popupList');

      userEvent.click(sortCaption);
      userEvent.click(bodyHeading);

      expect(list).not.toBeVisible();
    });
  });
});
