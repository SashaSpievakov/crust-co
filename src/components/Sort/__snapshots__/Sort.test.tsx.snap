// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sort Test renders the Sort component 1`] = `
Object {
  "children": Array [
    <div
      className="sc-gswNZR ekjJvS"
      onClick={[Function]}
    >
      <svg
        className="sc-hLBbgP bBGzuB"
        data-testid="popupIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 16 16"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M7.247 11.14 2.451 5.658C1.885 5.013 2.345 4 3.204 4h9.592a1 1 0 0 1 .753 1.659l-4.796 5.48a1 1 0 0 1-1.506 0z"
        />
      </svg>
      <b>
        Sort by
      </b>
      <span
        data-testid="currenChosenSortName"
      >
        rating
      </span>
    </div>,
    <section
      aria-hidden={true}
      className="sc-eDvSVe fQPTRR"
      open={false}
    >
      <ul
        aria-hidden={true}
        data-testid="popupList"
      >
        <li
          className="sc-jSUZER eGkRJW"
          onClick={[Function]}
        >
          rating
        </li>
        <li
          className="sc-jSUZER ebgEhB"
          onClick={[Function]}
        >
          price
        </li>
        <li
          className="sc-jSUZER ebgEhB"
          onClick={[Function]}
        >
          A to Z
        </li>
      </ul>
    </section>,
  ],
  "props": Object {
    "className": "sc-bcXHqe foXnjj",
  },
  "type": "section",
}
`;
